Firmware/MotorControl/trapTraj.cpp

```diff
@@ -0,0 +1,112 @@
+#include <math.h>
+#include "odrive_main.h"
+
+// Standard sign function, implemented to match the Python impelmentation
+template <typename T>
+int sign(T val) {
+    if (val == T(0))
+        return T(0);
+    else
+        return (std::signbit(val)) ? -1 : 1;
+}
+
+TrapezoidalTrajectory::TrapezoidalTrajectory(TrapTrajConfig_t &config) : config_(config) {}
+
+float TrapezoidalTrajectory::planTrapezoidal(float Xf, float Xi,
+                                             float Vi, float Vmax,
+                                             float Amax, float Dmax) {
+    float dx_stop = (Vi * Vi) / (Dmax * 2.0f);
+    float dX = Xf - Xi;
+    int s = sign(dX);
+
+    float Ar = s * Amax;          // Maximum Acceleration (signed)
+    float Dr = -1.0f * s * Dmax;  // Maximum Deceleration (signed)
+    float Vr = s * Vmax;          // Maximum Velocity (signed)
+
+    float Ta;
+    float Tv;
+    float Td;
+
+    // Checking for overshoot on "minimum stop"
+    if (std::abs(dX) <= dx_stop) {
+        Ta = 0;
+        Tv = 0;
+        Vr = Vi;
+        Dr = -1.0f * sign(Vi) * Dmax;
+        Td = std::abs(Vi) / Dmax;
+    } else {
+        // Handle the case where initial velocity > Max velocity
+        if ((s * Vi) > (s * Vr)) {
+            Ar = -1.0f * s * Amax;
+        }
+
+        Ta = (Vr - Vi) / Ar;  // Acceleration time
+        Td = (-Vr) / Dr;      // Deceleration time
+
+        // Peak Velocity handling
+        float dXmin = (Ta * (Vr + Vi)) / 2.0f + (Td * Vr / 2.0f);
+
+        // Short move handling
+        if (std::abs(dX) < std::abs(dXmin)) {
+            Vr = s * std::sqrt((-((Vi * Vi) / Ar) - 2.0f * dX) / ((1.0f / Dr) - (1.0f / Ar)));
+            Ta = std::max(0.0f, (Vr - Vi) / Ar);
+            Tv = 0;
+            Td = std::max(0.0f, -Vr / Dr);
+        } else {
+            Tv = (dX - dXmin) / Vr;
+        }
+    }
+
+    // Overshoot move handling
+    if(std::abs(Xf - Xi) <= ((Vi * Vi) / (2.0f * Dmax))) {
+        overshoot_move_ = true;
+        TrapTrajStep_t trajStep = evalTrapTraj(Ta+Tv+Td);
+        overshoot_time_ = Ta + Tv + Td;
+        return planTrapezoidal(Xf, trajStep.Y, trajStep.Yd, Vmax, Amax, Dmax); // Recalculate the move's second half
```

Comment:
Neater recursive style is to do the `Ta += overshoot_time_;` stuff up here after evaluating the lower trajectory "bare", instead of pushing down overshoot_move_ and overshoot_time_

Reply:
Not sure what you're referring to here.